### PR TITLE
Add support for send in server side

### PR DESCRIPTION
**Pull request by semhoun**
sbertin-telula patch for send in client was added to wakaama but not the server side.
So I fix the server side part.

### CODE REVIEW

File: core/observe.c [hunk 1]
```diff
@@ -1493,4 +1493,33 @@ bool observe_handleNotify(lwm2m_context_t * contextP,
...
+    for (clientP = contextP->clientList; clientP != NULL; clientP = clientP->next) {
+        if (clientP->sessionH == fromSessionH)
+            break;
+    }
+    if (clientP == NULL)
+        return COAP_400_BAD_REQUEST;
+
+    format = utils_convertMediaType(message->content_type);
+
+    if (contextP->monitorCallback != NULL) {
```
Comment (mlasch) [inline core/observe.c:1516]: I would rather create a separate callback with an interface like
```
void lwm2m_reporting_set_send_callback(lwm2m_context_t *contextP, lwm2m_result_callback_t callback, void *userData);
```
to report send messages. The `monitorCallback` should only be used for the "Client Registration Interface" and not for the "Information Reporting Interface". Also I would put this in a separate file like `core/reporting.c`.

File: core/internals.h [hunk 2]
```diff
@@ -455,6 +455,9 @@ void observe_clear(lwm2m_context_t * contextP, lwm2m_uri_t * uriP);
 bool observe_handleNotify(lwm2m_context_t * contextP, void * fromSessionH, coap_packet_t * message, coap_packet_t * response);
 void observe_remove(lwm2m_observation_t * observationP);
 lwm2m_observed_t * observe_findByUri(lwm2m_context_t * contextP, lwm2m_uri_t * uriP);
+#ifndef LWM2M_VERSION_1_0
```
Comment (mlasch) [inline core/internals.h:458]: It might be useful to add a check for `LWM2M_SERVER_MODE` in addition to `LWM2M_VERSION_1_0`.

File: core/observe.c [hunk 3]
```diff
@@ -1493,4 +1493,33 @@ bool observe_handleNotify(lwm2m_context_t * contextP,
     }
     return true;
 }
+
+#ifndef LWM2M_VERSION_1_0
```
Comment (mlasch) [inline core/observe.c:1497]: It might be useful to add a check for `LWM2M_SERVER_MODE` in addition to `LWM2M_VERSION_1_0`.

File: core/packet.c [hunk 4]
```diff
@@ -198,6 +198,11 @@ static uint8_t handle_request(lwm2m_context_t * contextP,
     case LWM2M_REQUEST_TYPE_REGISTRATION:
         result = registration_handleRequest(contextP, &uri, fromSessionH, message, response);
         break;
+#ifndef LWM2M_VERSION_1_0
```
Comment (mlasch) [inline core/packet.c:201]: It might be useful to add a check for `LWM2M_SERVER_MODE` in addition to `LWM2M_VERSION_1_0`.

**Review by mlasch — CHANGES_REQUESTED** [review]
Thanks for your contribution. This was still missing indeed, we have an internal patch for which we didn't find the time yet to upstream, but maybe it helps: https://github.com/husqvarnagroup/wakaama/tree/gardena/ml/server-side-send

File: core/packet.c (hunk 4 above)
Comment (semhoun) [reply core/packet.c:201]: Is already in a #ifdef LWM2M_SERVER_MODE (line 197)

Comment (mlasch) [reply core/packet.c:201]: Yes, my bad.

File: core/observe.c (hunk 1 above)
Comment (semhoun) [reply core/observe.c]: I made the change.

File: core/observe.c (hunk 3 above)
Comment (semhoun) [reply core/observe.c]: I made the change

**Comment by semhoun** [comment]
I made a separate callback as you suggested

**Comment by mlasch** [comment]
Nice, can you squash the "fix" commits and rebase your branch to the current upstream `main` branch, such that we get rid of the merge commits and have a nice linear commit history?

**Comment by semhoun** [comment]
I rebase to main. I squish all the commits, not only the fix because of the rebase.

**Review by mlasch — APPROVED** [review]

**Review by mlasch — APPROVED** [review]
Thanks for squashing. Looks good now.

**Pull request merged**